### PR TITLE
Add snapshot JDBC repository

### PR DESCRIPTION
**Pull request by MohamedFarouk-HMCTS**

### CODE REVIEW

File: event-sourcing/event-repository/event-repository-jdbc/src/main/java/uk/gov/justice/services/eventsourcing/repository/jdbc/snapshot/SnapshotJdbcRepository.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    private static final String JNDI_DS_EVENT_STORE_PATTERN = "java:/app/%s/DS.eventstore";
+
+
+    /**
+     * Insert the given aggregateSnapshot into Snapshot log.
+     *
+     * @param aggregateSnapshot the event to insert
+     * @throws uk.gov.justice.services.eventsourcing.common.exception.DuplicateSnapshotException if the
+     *                                                                                           version
+     *                                                                                           already
```
Comment (jaceko) [inline event-sourcing/event-repository/event-repository-jdbc/src/main/java/uk/gov/justice/services/eventsourcing/repository/jdbc/snapshot/SnapshotJdbcRepository.java]: a bit strange formatting

File: event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package uk.gov.justice.services.eventsourcing.common.snapshot;
+
+import java.util.UUID;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+/**
```
Comment (jaceko) [inline event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java]: we don't put author comments in the framework

File: messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                        .withClientCorrelationId("clientId")
+                )
+                .withPayloadOf("value_1", "object", "name_1")
+                .withPayloadOf("value_2", "object", "name_2")
+                .withPayloadOf("value_3", "name_3")
+                .build();
+
+
+        System.out.println(envelope.toDebugStringPrettyPrint());
+
```
Comment (jaceko) [inline messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java]: this test does not test anything (no assertions)

File: messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.justice.services.messaging;
+
+import static java.util.UUID.randomUUID;
+import static uk.gov.justice.services.messaging.DefaultJsonEnvelope.envelope;
+import static uk.gov.justice.services.messaging.JsonObjectMetadata.metadataOf;
+
+import org.junit.Test;
+
+public class AllanTest {
```
Comment (jaceko) [inline messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java]: allan test?

File: event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public int hashCode() {
+        return new HashCodeBuilder(17, 37)
+                .append(streamId)
+                .append(sequenceId)
+                .append(type)
+                .append(aggregate)
+                .toHashCode();
+    }
```
Comment (jaceko) [inline event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java]: we should probably have unit tests for the equals and hashcode methods

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/snapshot/SnapshotRepositoryJdbcIT.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+        Optional<AggregateSnapshot> snapshot = jdbcRepository.getLatestSnapshot(STREAM_ID);
+
+        assertThat(snapshot.isPresent(), is(false));
+    }
+
+    @Test
+    public void shouldThrowDuplicateSnapshotException() throws DuplicateSnapshotException, InvalidSequenceIdException {
+        expectedException.expect(isA(JdbcRepositoryException.class));
+        expectedException.expectCause(isA(org.h2.jdbc.JdbcSQLException.class));
```
Comment (jaceko) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/snapshot/SnapshotRepositoryJdbcIT.java]: we shouldn't be using H2 specific exceptions

File: event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java (hunk 2 above)
Comment (MohamedFarouk-HMCTS) [inline event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java]: Done

File: messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java (hunk 3 above)
Comment (MohamedFarouk-HMCTS) [inline messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java]: Removed as it is invalid class

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/snapshot/SnapshotRepositoryJdbcIT.java (hunk 6 above)
Comment (MohamedFarouk-HMCTS) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/snapshot/SnapshotRepositoryJdbcIT.java]: removed

File: event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java (hunk 5 above)
Comment (MohamedFarouk-HMCTS) [inline event-sourcing/event-sourcing-common/src/main/java/uk/gov/justice/services/eventsourcing/common/snapshot/AggregateSnapshot.java]: Added EqualsTester unit tests

File: messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java (hunk 4 above)
Comment (MohamedFarouk-HMCTS) [inline messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/AllanTest.java]: Removed as it should not be in this module at all

**Pull request merged**